### ISSUE TITLE
Least Squares Optimization Fails for x0 << 1.0.

### ISSUE THREAD

**Issue by tomfallen** [scipy/scipy#10767]
The scipy.optimize.least_squares fails to minimize a well behaved function when given starting values much less than 1.0.  In particular, I'm considering the function `f(x) = x - 3.0`.  If `x0 = 0.0` it optimizes well, but `x0 = 1e-9` (or anything smaller but non-zero) it doesn't move.

The root cause seems to be a numerical issues in the underlying MINPACK Fortran code.  When calculating the Jacobian it uses a step size `h`, where `h = abs(x0) * eps`, and `eps` is related to machine precision, in this case about `1.5e-8`.  For small `h`, `h * eps` ends up very small.  The difference between `x0` and `x0 + h`  gets truncated after subtracting 3.0, so `f(x)` and `f(x +h)` are equal, the Jacobian is zero, and the optimization doesn't move.

#### Reproducing code example:
```
from scipy.optimize import least_squares

X_GOAL = 3.0

def myfunc(x):
    return x - X_GOAL

for x0 in [0.0, 1e-9]:
    res = least_squares(myfunc, x0, method='lm', verbose=2)
    x_opt = res['x']
    print("x0:", x0, "x_opt:", x_opt, "x_error", x_opt - X_GOAL)
```
This will show an error of zero for `x0 = 0.0`, and an error of `3.0` for `x0 = 1e-9`.  This seems like undesirable behavior, and is causing problems for me in a more complicated real-life example.

I haven't been able to find any settings that resolve this issues.  Using a very large `diff_step` should get the first iteration to move, but that value will be far too large as soon as `x` is not very small.

I took a look at the Fortran code, and there are two places where there's a check for a value being exactly zero then sets it to some value.  Changing these to saturate at the specified value resolves this particular issue for me.  Saturation at small values is also the approach taken in `approx_derivative` in `scipy/optimize/_numdiff.py`, where the absolute step size is calculated as `rel_step * sign_x0 * np.maximum(1.0, np.abs(x0))` (in `_compute_absolute_step`). 

The behavior with these changes seems more desirable than the current behavior, but changing historic behavior seems a bit concerning.

Here are the specific changes:

```
scipy/scipy/optimize/minpack/fdjac2.f, line 94
- if (h .eq. zero) h = eps
+ h = dmax1(h, eps)
```
```
scipy/optimize/minpack/lmder.f, line 272
- if (delta .eq. zero) delta = factor
+ delta = dmax1(delta, factor)
```

With these two changes all the unit tests pass, except for  `test_diff_step` in `scipy/optimize/tests/test_least_squares.py`, which requires a much smaller `diff_step` to behave properly.  This is using the 'lm' solver. I believe the 'trf' and 'dogbox' methods utilize different code but exhibit the same behavior.

#### Scipy/Numpy/Python version information:
```
1.3.1 1.17.0
sys.version_info(major=3, minor=5, micro=2, releaselevel='final', serial=0)
```

**Comment by andyfaff** [comment]
@nmayorov, I think you'd be best placed to answer this. I'd be for improved behaviour in this circumstance, I'm not so sure exact (bitwise) back compatibility is as important.

**Comment by nmayorov** [comment]
I will investigate it.

**Comment by nmayorov** [comment]
This behavior is not due to numerical differentiation, the Jacobian is estimated correctly as 1.

All optimization methods in `least_squares` are trust-region methods. At the beginning the initial radius of the trust region is selected as 
```
    Delta = norm(x0 * scale_inv)
    if Delta == 0:
        Delta = 1.0
```
By default `scale_inv` is 1. So the idea behind this is to use `x0` as the scale of how much variables might change during the optimization process. If `x0 = 1000` then we expect it to change within `1000` approximately. This makes sense for sure.

So when you set `x0=1e-9` the algorithm assumes that your variables needs to be optimized within this range. The first step is `1e-9` and the cost function change is sufficiently small for algorithm to terminate.

I'd put this behavior into poor starting point selection category. I guess it can be fixed (sort of), but I don't see it improving the actual quality of the solver. That is real issues on hard problems are not related to this little quirk at all.

**Comment by tomfallen** [comment]
@nmayorov Thanks for taking a look at this.

I'm looking at the  Levenberg-Marquardt method, and it doesn't appear that Jacobian in correctly estimated.  If I print out `fjac(i,j)` (in scipy/optimize/minpack/fdjac2.f near line 100), I get zero.  The calculation of `delta` in `scipy/optimize/minpack/lmdif.f` is also a problem, but only after I modify the Jacobian calculation.

I hear what you're saying about poor starting point selection, but that doesn't help in the actual problem I have.  I'm optimizing a function that changes somewhat, using the optimal state from the previous step as the starting point of the next optimization, so I don't have much control over the starting state.  Sometimes the optimum is actually zero, so the approximately optimal solution ends up very small, which is then used in the next step, and the optimization gets stuck.  I can put in a hack where I check for the  optimization not moving along with near zero start state, then manually reset the start state to exactly zero and re-optimize, but this isn't ideal.  Any suggestions would be welcome.

It also seems like there's potential to mess up optimization problems that don't start with very small values of x0.  With each iteration of the linear problem there's a new x0 produced by the previous linear problem, right?  If this x0 happens to be near zero, couldn't the optimization stop prematurely?

I believe that my suggestion of lower-bounding both `h` and `delta` would fix the problem with small starting state when the optimal state is not small, and that the trade-off would be lower precision for problems in which the state being optimized is very small.  However, optimizing a problem where the state is in the range of, say, `1e-6`, isn't really best practices anyway, and I believe there'd be no drawback from problems which are appropriately normalized.

**Comment by nmayorov** [comment]
OK, let's try to figure out what is going on and what can be changed.

> I'm looking at the Levenberg-Marquardt method, and it doesn't appear that Jacobian in correctly estimated. If I print out fjac(i,j) (in scipy/optimize/minpack/fdjac2.f near line 100), I get zero. The calculation of delta in scipy/optimize/minpack/lmdif.f is also a problem, but only after I modify the Jacobian calculation.

The code which selects the perturbations step size in MINPACK (you have probably seen it):
```
         temp = x(j)
         h = eps*dabs(temp)
         if (h .eq. zero) h = eps
```
So for `x0 = 1e-9` we get `h ~ 1e-25` and `fun(x0 + h) - fun(x0) = 0`. This again conceptually explained by a poor initial point selection. In `approx_derivative` we have `h = max(1, abs(x0)) * eps` so this part sort of works. 

But there is no perfect or right solution for numerical differentiation. So I'm overall against changing this part (or anything at all) in MINPACK, because it doesn't fundamentally improve it and it's better to keep the canonical Fortran implementation. 

> I'm optimizing a function that changes somewhat, using the optimal state from the previous step as the starting point of the next optimization, so I don't have much control over the starting state. Sometimes the optimum is actually zero, so the approximately optimal solution ends up very small, which is then used in the next step, and the optimization gets stuck. I can put in a hack where I check for the optimization not moving along with near zero start state, then manually reset the start state to exactly zero and re-optimize, but this isn't ideal. Any suggestions would be welcome.

But it still looks to me as a poor initial point selection. This basic "continuation" strategy is generally good, but doesn't always work, as you say. You may somehow adjust it, i.e. move start point to 1 (or other appropriate value corresponding to scale of your variables) if it's close to 0.

> It also seems like there's potential to mess up optimization problems that don't start with very small values of x0. With each iteration of the linear problem there's a new x0 produced by the previous linear problem, right? If this x0 happens to be near zero, couldn't the optimization stop prematurely?

Sorry, but didn't catch the idea here. Could you rephrase or clarify?

> I believe that my suggestion of lower-bounding both h and delta would fix the problem with small starting state when the optimal state is not small, and that the trade-off would be lower precision for problems in which the state being optimized is very small. However, optimizing a problem where the state is in the range of, say, 1e-6, isn't really best practices anyway, and I believe there'd be no drawback from problems which are appropriately normalized.

In `trf` and `dogbox` we have `h` lower-bounded, in MINPACK, as I said, I'm hesitant to change it.

For `delta` --- I believe lower-bounding it is not a great idea, however providing an arguments to control its initial values is a decent idea.

MINPACK supports it in the form 
```
factor: float, optional
  A parameter determining the initial step bound (factor * || diag * x||). Should be in interval (0.1, 
  100).
```
For new methods I'd simplify this logic to directly setting the initial radius of trust regions (I believe this is the approach in ceres-solver C++ library).

Another option I have in mind is to require that the trust regions doesn't restrict the step in order to declare the convergence by `ftol` or `xtol`. I think it is worth checking.

**Comment by tomfallen** [comment]
> Sorry, but didn't catch the idea here. Could you rephrase or clarify?

My thought here is that the optimization could accidentally land on an intermediate state that is small, leading to small `delta` and `h`, which could cause premature stopping.  Say we're solving a problem where the optimum is at  `x =1.0`, and we start at `x = -1.0`.  In the first step we calculate the Jacobian and take a step of, say `1.0 + 1e-16`  In the next iteration `x`  is now `1e-16`.  This then runs into the same problem I'm seeing, where the small starting state results in very small `delta` and `h`, causing the optimization to terminate at `1e-16` rather than near `1.0`.  I believe this premature stopping is possible even if the starting state, optimal state, and function are all on the order of `1.0`.

It also seems strange to me to have discontinuous behavior at `x0 = 0.0`.  As noted above, `approx_derivative` in `scipy/optimize/_numdiff.py` scales the finite difference step size with `max(abs(x0), 1.0)` which makes more sense to me.

It seems to me that the current behavior could also result in a problem somewhat opposite to the one I'm having.  If I have a problem whose state is order `1e-9` then starting with `x0 = 0.0` (which seems pretty natural) would result in both `delta` and `h` being too large, but starting with `x0 = 1e-13` would work well.  This seems pretty intuitive to me.  And if you are looking to optimize problems with very small state, it seems like rescaling the problem to be order `1.0` would be a better approach than choosing small (but not exactly 0.0) starting state.

I like the idea of using `xtol` and/or `ftol` as a guide, either to prevent termination or to set a floor on the size of the trust region and finite difference step.  I'm not seeing an obvious way to do this without modifying the Fortran code, but I'm also having a little troubles following that code.  I'd also be fine with a manual setting to lower bound both `delta` and `h`, but I'm not seeing a way to do that either.

**Comment by tomfallen** [comment]
I'm not seeing an approach that would fix my particular issue without modifying the original fortran code.  The numeric derivatives in the LM solver are evaluating to zero (due to very small h while the functions output is on the order of 1), and there doesn't appear to be a way to set a floor on the size of h, other than using exactly zero, which the LM solver treats as if it were order 1.  I have a hack where I check my initial state and set any elements that are smaller than some threshold to exactly zero.  It's a bit unsatisfactory, but seems to work, so I think I'll go with that for now.  Thanks for the help.

**Comment by nmayorov** [comment]
@tomfallen 

Feel free to report other issues or reopen this. Closing for now.
